### PR TITLE
Adding ability to pipe results into payloads using JSONpaths

### PR DESCRIPTION
**Pull request by mshick**
This PR adds the ability to process JSONPath strings in a batch request payload. So, `$..[0].name` would match a `name` prop on the first object in the results array. Paths that resolve to undefined set undefined on the resulting payload, and paths at the root will return an empty object due to how an undefined payload is evaluated in bassmaster.

Example:

```js
[
{ method: 'get', path: '/item' },
{ method: 'post', path: '/echo', payload: '$..[0]' }
]
```

or this

```js
[
{ method: 'get', path: '/item' },
{ method: 'post', path: '/echo', payload: { data: '$..[0].name' } }
]
```

A few notes:

* I've decided to use a [JSONPath](https://github.com/s3u/JSONPath) string, in conjunction with the JSONPath dependency, which is applied to the resultsMap. After initially attempting to support the unique bassmaster syntax used for paths, e.g., `$[index]`, I figured it was more consistent with practices in other batch endpoints (Facebook) to stick to JSONpaths only.

* [traverse](https://github.com/substack/js-traverse) is used to walk the payload objects and test for potential jsonpaths, matching `$..` at the beginning of the string. Since the jsonpath is being applied to an array, this is the only valid form that can / should be used.

* tests have been added, and all are passing with coverage at 100%

### CODE REVIEW

**Comment by cadecairos** [comment]
This is pretty cool. It will need some rebasing

**Comment by cadecairos** [comment]
I think we're going to want to use Hoek.reach VS JSONPath

see #56

File: lib/batch.js [hunk 1]
```diff
@@ -154,18 +161,15 @@ internals.batch = function (batchRequest, resultsData, pos, parts, callback) {
                         path += value;
                     }
                     else {
-                        error = new Error('Reference value includes illegal characters');
```
Comment (cadecairos) [inline lib/batch.js]: Lets not change the error handling here to use try/catch. It really should only ever be used if parsing JSON.

**Comment by mshick** [comment]
Hey hey. I'm really sorry for the wait here. Been traveling and let this slip.

Hopefully this change has things in order?

**Comment by devinivy** [comment]
Looking forward to this– nice work! :+1:

**Comment by geek** [comment]
@cadecairos any update on this?  It looks good to me, mind if I merge and cleanup and publish as a 1.9.0 ?

**Comment by mshick** [comment]
@geek thanks, yeah, this has been baking for awhile! Can't even remember what I was building when I submitted the PR.

**Comment by cadecairos** [comment]
Yeah, sorry for the delay. I'm out with my family today, but I'll take care of this next time I'm online.

**Pull request merged**